Overhaul OxyPlot.ImageSharp (#1530)

VisualMelon opened the pull request:
Addresses #1530 

### Checklist

- [x] I have included examples or tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [x] I have cleaned up the commit history

### Changes proposed in this pull request:

- Change OxyPlot.ImageSharp to target .NET Standard 1.3, bringing it inline with the base library
- Add `DrawImage` support to `OxyPlot.ImageSharp.PngRenderContext`
- Add Dpi awareness to `OxyPlot.ImageSharp.PngRenderContext` and `OxyPlot.ImageSharp.PngExporter`
- Add rotated text support to `Oxyplot.ImageSharp`
- Makes `Oxyplot.ImageSharp` font sizes conform to 96DPI standard
- Add clipping support to `Oxyplot.ImageSharp`
- Add dashed line support to `Oxyplot.ImageSharp`
- (Code breaking) Add a constructor to `OxyPlot.ImageSharp.PngExporter`, meaning there is no longer a default constructor
- Add basic tests for the `PngExporter` copied from `OxyPlot.SkiaSharp.Tests`
- Add a `JpegExporter` along with tests

Dpi awareness is based directly on the implementation for `OxyPlot.SkiaSharp`, and as such would appreciate a review by @Jonarw. There is a lot of shared logic, but type discrepancy mean there is relatively little shared code, and none of it is very interesting; it may be worth providing helper methods (within `RenderContextBase`?) for Dpi awareness to reduce bloat and ensure consistency between platforms going forward. The change undoes a hack to do with a now result issue with dpi aware text rendering in ImageSharp.

I am putting together [some notes](https://gist.github.com/VisualMelon/d1eb4bca4f52de86353421f96bcaeb58) on the implementation to explain why it is the way it is, so that when someone looks at it, proclaims "That's hideous!", and decides to rewrite it, they will hopefully know more going into it then I did.

There are a couple of TODOs left, but they are small things that won't have sweeping design consequences.

@oxyplot/admins

VisualMelon commented:
I can't see why snyk is failing, so I'm not sure how I can address that at this time.

objorke commented:
I am testing the snyk tool, was not aware that it was added as a required check here. Can remove it if it blocks the PR!

VisualMelon commented:
@objorke do you have any idea why Snyk is unhappy? It seems I don't have authorisation to view the report.

objorke commented:
Snyk says "Expected target framework netstandard2.0 is not present in proj file" with link to https://github.com/oxyplot/oxyplot/blob/develop/Source/OxyPlot.ImageSharp/OxyPlot.ImageSharp.csproj
To me this looks like a false positive error report.

objorke commented:
I have deactivated the Snyk webhook. I guess we need a change here to trig the checks again.

VisualMelon commented:
@objorke currently rewriting `DrawImage`, so should hopefully have a change soon...

VisualMelon commented:
The `DrawImage` method should now approximate the `SkiaSharp` renderer. I've added additional tests to both in order to verify this.

There is significantly complexity in the `DrawImage` method in order to provide sub-pixel resolution and edge-clamped sampling. I'm steadily coming to the conclusion it would be better to implement a more advanced `DrawImage` method in `ImageSharp` itself for us to use, rather than try to work around the existing API; however, this seems to work well enough at the moment.

Once someone has sanity checked the changes, I'll probably squash everything into one commit, because it's all relevant to #1530, though the additional SkiaSharp tests are pushing that a bit.

Jonarw commented on Source/OxyPlot.ImageSharp/OxyPlot.ImageSharp.csproj:
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
     <PropertyGroup>
-        <TargetFramework>netstandard2.0</TargetFramework>
+        <TargetFramework>netstandard1.3</TargetFramework>
```
Should we add `<LangVersion>8</LangVersion>`? Then we could use those tasty switch expressions ;)

VisualMelon replied:
Personally I deplore switch expressions, but scoped using is nice yo have ;)

VisualMelon replied:
I caved: langversion 8 it shall be ;)

Jonarw commented on Source/OxyPlot.ImageSharp/PngRenderingContext.cs:
```diff
@@ -75,42 +87,286 @@ public void Save(Stream output)
+        }
+
         /// <inheritdoc/>
         public override void DrawText(ScreenPoint p, string text, OxyColor fill, string fontFamily = null, double fontSize = 10, double fontWeight = 400, double rotation = 0, OxyPlot.HorizontalAlignment horizontalAlignment = OxyPlot.HorizontalAlignment.Left, OxyPlot.VerticalAlignment verticalAlignment = OxyPlot.VerticalAlignment.Top, OxySize? maxSize = null)
```
It would be nice to support text rotation, but from a quick search ImageSharp does not seem to support that... Or is there a way?

VisualMelon replied:
I've done rotated text with ImageSharp before, but it was only mutliples of 90 degrees. I'll dig out that old code and see what I did there - because I feel like I was using transforms in a way that my recent experience suggests don't exist - and get back to you.

VisualMelon replied:
Nope: that old code is just rotating the whole image. I can't shake the feeling that I must be missing something about ImageSharp, because everything just feels wrong.

As I workaround, we could render the text to another image, rotate that image, and then draw it on top.

Jonarw commented on Source/OxyPlot.ImageSharp/PngRenderingContext.cs:
```diff
@@ -119,27 +375,154 @@ public override OxySize MeasureText(string text, string fontFamily = null, doubl
+                var pen = actualDashArray != null
+                    ? new Pen(ToRgba32(stroke), actualThickness, actualDashArray)
+                    : new Pen(ToRgba32(stroke), actualThickness);
+                var actualPoints = this.GetActualPoints(points, thickness, edgeRenderingMode).ToArray();
```
I think you have to pass the `actualThickness` here.

VisualMelon replied:
Actually, `GetSnapOffset` inside `GetActualPoints` performs the transform itself. This is consistent with `SkiaSharp`

Jonarw replied:
Oh ok sorry, my mistake then.

Jonarw commented on Source/OxyPlot.ImageSharp/PngRenderingContext.cs:
```diff
@@ -148,14 +531,21 @@ public override void DrawPolygon(IList<ScreenPoint> points, OxyColor fill, OxyCo
+                var pen = actualDashArray != null
+                    ? new Pen(ToRgba32(stroke), actualThickness, actualDashArray)
+                    : new Pen(ToRgba32(stroke), actualThickness);
+                var actualPoints = this.GetActualPoints(points, thickness, edgeRenderingMode).ToArray();
```
Also `actualThickness`.

VisualMelon replied:
(as above)

VisualMelon commented:
@Jonarw yeah, just scaling and snapping is what I had in mind. I won't do that as part of this, because we may want to incorporate it into other exporters, and this is PR is already getting out of hand.

Jonarw commented on Source/OxyPlot.ImageSharp/PngRenderingContext.cs:
```diff
@@ -75,42 +87,286 @@ public void Save(Stream output)
+        /// <param name="strokeThickness">The stroke thickness.</param>
+        /// <param name="edgeRenderingMode">The edge rendering mode.</param>
+        /// <returns>The actual rectangle.</returns>
+        private RectangleF GetActualRect(OxyRect rect, double strokeThickness, EdgeRenderingMode edgeRenderingMode)
+        {
+            switch (edgeRenderingMode)
+            {
+                case EdgeRenderingMode.PreferGeometricAccuracy:
+                case EdgeRenderingMode.PreferSpeed:
+                    return this.Convert(rect);
+                default:
+                    var actualThickness = this.GetActualThickness(strokeThickness, edgeRenderingMode);
+                    var snapOffset = GetSnapOffset(actualThickness);
+                    return this.ConvertSnap(rect, snapOffset);
+            }
+        }
+
+        /// <summary>
+        /// Gets the <see cref="RectangleF"/>s that should actually be rendered from the list of <see cref="OxyRect"/>s, taking into account DPI scaling and snapping if necessary.
+        /// </summary>
+        /// <param name="rects">The rectangles.</param>
+        /// <param name="strokeThickness">The stroke thickness.</param>
+        /// <param name="edgeRenderingMode">The edge rendering mode.</param>
+        /// <returns>The actual rectangles.</returns>
+        private IEnumerable<RectangleF> GetActualRects(IEnumerable<OxyRect> rects, double strokeThickness, EdgeRenderingMode edgeRenderingMode)
+        {
+            switch (edgeRenderingMode)
+            {
+                case EdgeRenderingMode.PreferGeometricAccuracy:
+                case EdgeRenderingMode.PreferSpeed:
+                    return rects.Select(this.Convert);
+                default:
+                    var actualThickness = this.GetActualThickness(strokeThickness, edgeRenderingMode);
+                    var snapOffset = GetSnapOffset(actualThickness);
+                    return rects.Select(rect => this.ConvertSnap(rect, snapOffset));
+            }
+        }
```
These appear to be unused, so can be removed.

VisualMelon replied:
Should probably override `DrawRectangles` at some point, but for now I will just remove these.

Jonarw commented:
A minor general point: It would be nice to have consistent use of `this` and `var`. I think most portions of the library use `this` and `var` wherever possible, so it would be a nice opportunity to bring the `PngRenderingContext` in line while we are modifying most of its code anyway.

Another small-ish point:
Missing rotation aside - vertical text alignment also seems slightly off:
![TextAnnotation - TextAnnotation - Transposed](https://user-images.githubusercontent.com/18097859/80974943-bf229100-8e21-11ea-8b82-5ff5bdc0cda3.png)
(notice that the 'Text annotation 1' text is a bit high in the box).
But we could also leave this for a future issue, this is already a huge improvement for the `OxyPlot.ImageSharp` library!

VisualMelon commented:
@Jonarw yep: sorting out `this` should be done in latests commits (though the jpeg stuff has gone wrong, so don't look at that); I'll check `var`, and do a StypeCop pass at some point.

Not sure what to make of the vertical text alignment... might just be ImageSharp's own take on fonts, but I'll double check it's not me being silly.

Thanks for your prompt feedback.

Jonarw commented:
> Not sure what to make of the vertical text alignment... might just be ImageSharp's own take on fonts, but I'll double check it's not me being silly.

In `SkiaRenderContext` I ended up just using the Ascent and Descent font metrics for vertical alignment to make text baselines consistent across runs of different text - but I don't know if you have these metrics available in ImageSharp.

VisualMelon commented:
@Jonarw I guess `Font.Ascender` and `Font.Descender` are the same thing: I'll take a good look at your code then and try to sort it out.

VisualMelon commented:
Slowly getting somewhere...

![TextAnnotation - TextAnnotation - Transposed](https://user-images.githubusercontent.com/5646475/81013683-797ebc00-8e53-11ea-829d-7c63b7cee46c.png)

VisualMelon commented:
What feels like a week later, I have text rendering in a vaguely stable state.

Skia
![TextAnnotation - TextAnnotation - Transposed](https://user-images.githubusercontent.com/5646475/81055270-41fa2900-8ec0-11ea-819b-1efd32584cdb.png)

ImageSharp
![TextAnnotation - TextAnnotation - Transposed](https://user-images.githubusercontent.com/5646475/81055251-3870c100-8ec0-11ea-9465-899e61184580.png)

The font size discrepency is soewhat disconcerting, and I'll be looking into that next but the real world may get in the way.

P.S. SkiaSharp doesn't seem to support multiline text. I've implemented it for ImageSharp, but not properly tested.

VisualMelon commented:
@objorke do you recall what the `FontScaleFactor` is achieving in the `GraphicsRenderContext`?

Comparing outputs to work, I think the WinForms and Skia renders have somewhat small text, but they just about match Word and the ImageSharp render when I make the following changes:
 - Change `GraphicsRenderContext` such that `FontScaleFactor = 1.0` (rather than `0.8`)
 - Change `SkiaRenderContext.GetTextPaint` such that `TextSize = this.Convert(fontSize / 0.75);` (convert points (1/72nds of an inch) to 1/96ths of an inch (nominal dpi))

@Jonarw does this make sense to you? I can't find any documentation for what `TextSize` really is (based on your code and these observations I am assuming it is pixels)

Jonarw commented:
> @Jonarw does this make sense to you? I can't find any documentation for what TextSize really is (based on your code and these observations I am assuming it is pixels)

With the SkiaSharp renderer I took the WPF `CanvasRenderContext` as a reference. On WPF text size is specified in 1/96 inch, so exactly 1 pixel at 100% UI scaling.

In SkiaSharp the text size is specified in pixels, and as neither SkiaSharp nor WPF do any scaling of the font size at the moment they give similar output.

In the end we should just define what unit the text size is, put that in the documentation and stick to it everywhere. Considering that WPF is probably the most used platform of oxyplot, should we define the unit of text size to 1/96 inch? If ImageSharp uses 1/72 inch, you would have to multiply font sizes by 0.75.

> P.S. SkiaSharp doesn't seem to support multiline text. I've implemented it for ImageSharp, but not properly tested.

Yes I also realized this recently, I plan to address that in the near future.

VisualMelon commented:
@Jonarw it would seem odd to use something other than points, but not knowing the history and the fact that it's been this way for a while, I'm hesitant to make a suggestion to change. If everything else is going for 1/96ths, then I'll change ImageSharp to agree for now. As you say, this needs documenting somewhere...

Should the WinForms scale factor be changed to 0.75 as well?

objorke commented:
yes, this library was first implemented for WPF, and simply used WPF device-independent units (1/96 inch) for font sizes, line thickness etc.

VisualMelon commented:
@objorke thanks for clarifying!

VisualMelon commented:
Rebased, squashed, and style-cop'd. There are a couple of TODOs in there, but they shouldn't motivate any changes in the design. I don't intend to do much more for this already bloated PR.

Most of this is truly horrendous code, and I'm putting a few [notes in a gist](https://gist.github.com/VisualMelon/d1eb4bca4f52de86353421f96bcaeb58) to go some way to explaining why this is.

Jonarw commented:
Wow, respect for going through all that trouble!
I can't claim that I've gone through every line of the code, but I have not seen any obvious issues.
Test output also seems fine, so to me this looks good!

VisualMelon commented:
(Just rebased to resolve merge conflict in changelog)

objorke commented:
Fixed the snyk check

VisualMelon commented:
@objorke thanks. I'm going to divert my effors to #1531 now that this font stuff is fresh in my mind. Is there anything else you wanted done with this, or is it mergeable? When I run out of fun things to do I might come back and try to tidy this a little, maybe looking at reducing redundancy between Dpi aware renders while I am at as (as in the discussion somewhere above).

objorke commented:
This looks good to me, but I am not able to follow all the details... 
I am very happy to see all the development/improvements of this project right now!

VisualMelon commented:
Merged. Feel free to shout at me if anything ImageSharp related goes wrong.
